src/main/java/org/jenkinsci/remoting/engine/jnlp3/ChannelCiphers.java

```diff
@@ -0,0 +1,100 @@
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+ * THE SOFTWARE.
+ */
+package org.jenkinsci.remoting.engine.jnlp3;
+
+import javax.crypto.Cipher;
+import javax.crypto.SecretKey;
+import javax.crypto.spec.IvParameterSpec;
+import javax.crypto.spec.SecretKeySpec;
+import java.io.IOException;
+
+/**
+ * {@link javax.crypto.Cipher}s that will be used to construct an encrypted
+ * {@link hudson.remoting.Channel} after a successful handshake.
+ *
+ * @author Akshay Dayal
+ */
+public class ChannelCiphers {
+
+    private final byte[] aesKey;
+    private final byte[] specKey;
+    private final Cipher encryptCipher;
+    private final Cipher decryptCipher;
+
+    ChannelCiphers(byte[] aesKey, byte[] specKey, Cipher encryptCipher, Cipher decryptCipher) {
+        this.aesKey = aesKey;
+        this.specKey = specKey;
+        this.encryptCipher = encryptCipher;
+        this.decryptCipher = decryptCipher;
+    }
+
+    public byte[] getAesKey() {
+        return aesKey;
+    }
+
+    public byte[] getSpecKey() {
+        return specKey;
+    }
+
+    public Cipher getEncryptCipher() {
+        return encryptCipher;
+    }
+
+    public Cipher getDecryptCipher() {
+        return decryptCipher;
+    }
+
+    /**
+     * Create a pair of AES symmetric key {@link javax.crypto.Cipher}s using
+     * randomly generated keys.
+     *
+     * @throws IOException If there is a problem constructing the ciphers.
+     */
+    public static ChannelCiphers create() throws IOException {
+        return create(Jnlp3Util.generate128BitKey(), Jnlp3Util.generate128BitKey());
+    }
+
+    /**
+     * Creates a pair of AES symmetric key {@link javax.crypto.Cipher}s using
+     * the given AES key and {@link javax.crypto.spec.IvParameterSpec} key.
+     *
+     * @throws IOException If there is a problem constructing the ciphers.
+     */
+    public static ChannelCiphers create(byte[] aesKey, byte[] specKey) throws IOException {
+        try {
+            SecretKey secretKey = new SecretKeySpec(aesKey, "AES");
+            IvParameterSpec spec = new IvParameterSpec(specKey);
+            Cipher encryptCipher = Cipher.getInstance(CIPHER_TRANSFORMATION);
+            encryptCipher.init(Cipher.ENCRYPT_MODE, secretKey, spec);
+            Cipher decryptCipher = Cipher.getInstance(CIPHER_TRANSFORMATION);
+            decryptCipher.init(Cipher.DECRYPT_MODE, secretKey, spec);
+
+            return new ChannelCiphers(aesKey, specKey, encryptCipher, decryptCipher);
+        } catch (Exception e) {
+            throw new IOException("Failed to create channel ciphers", e);
```

Comment:
I feel a bit uneasy about ChannelCiphers and HandshakeCiphers throwing IOExceptions. I feel they should throw an Exception and JnlpProtocol.establishChannel should be allowed to bubble up an Exception to Engine, which can be caught and handled.